tunix/rl/agentic/TrajectoryCollector/trajectory_collect_engine.py

```diff
@@ -0,0 +1,135 @@
+    # --------------------------------------------------
+    def __init__(
+        self,
+        agent: BaseAgent,
+        env: BaseEnv,
+        *,
+        model_call: Callable[[list[Dict[str, str]]], str],
+        final_reward_fn: Optional[Callable[[Dict, str], float]] = None,
+        max_steps: int = 10,
+        gamma: float = 1.0,
+        timeout: float = 30.0,
+    ):
+        self.agent = agent
+        self.env = env
+        self.model_call = model_call
+        self.final_reward_fn = final_reward_fn or (lambda *_: 0.0)
+        self.max_steps = max_steps
+        self.gamma = gamma
+        self.timeout = timeout
+
+    # --------------------------------------------------
+    # Public API
+    # --------------------------------------------------
+    async def collect(self) -> Trajectory:
+        """Performs a full rollout and returns the agent's trajectory"""
+        await self._reset()
+        for t in range(self.max_steps):
+            done = await self._one_step(t)
+            if done:
+                break
+        await self._append_final_reward()
+        self._fill_returns()
+        await self._close()
+        return self.agent.trajectory
+
+    @staticmethod
+    async def collect_many(
+        pairs: List[Tuple[BaseAgent, BaseEnv]],
+        *,
+        model_call: Callable[[list[Dict[str, str]]], str],
+        final_reward_fn: Optional[Callable[[Dict, str], float]] = None,
+        max_steps: int = 10,
+        gamma: float = 1.0,
+        timeout: float = 30.0,
+    ) -> AsyncGenerator[Tuple[int, Trajectory], None]:
+        """
+        Run multiple (agent, env) pairs concurrently.
+        Yields results one by one as they complete.
+
+        Yields:
+            (index, trajectory)
+        """
+
+        async def _run_one(i: int, agent: BaseAgent, env: BaseEnv):
+            engine = TrajectoryCollectEngine(
+                agent, env,
+                model_call=model_call,
+                final_reward_fn=final_reward_fn,
+                max_steps=max_steps,
+                gamma=gamma,
+                timeout=timeout,
+            )
+            traj = await engine.collect()
+            return i, traj
+
+        tasks = [_run_one(i, a, e) for i, (a, e) in enumerate(pairs)]
+        for coro in asyncio.as_completed(tasks):
+            yield await coro
+
+    # --------------------------------------------------
+    # Internal helpers
+    # --------------------------------------------------
+    async def _reset(self):
+        obs, _ = await asyncio.get_event_loop().run_in_executor(None, self.env.reset)
+        self.agent.reset()
+        self.agent.update_from_env(obs, 0.0, False, {})
+        self._start_ts = time.time()
+
+    async def _one_step(self) -> bool:
+        # 1) Call LLM
+        resp = await asyncio.get_event_loop().run_in_executor(
+            None, self.model_call, self.agent.chat_completions
+        )
+        action = self.agent.update_from_model(resp).action
+
```

Comment:
do you need to add some check of action here? for example, it should not be None?